## Nivell 1

**Exercici 
1:**
Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un pandas Dataframe. Explora les dades que conté, i queda’t únicament amb les columnes que consideris rellevants.

In [1]:
import pandas as pd
import seaborn as sns

In [2]:
df=pd.DataFrame(pd.read_csv('DelayedFlights.csv'))

In [3]:
df.columns

Index(['Unnamed: 0', 'Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime',
       'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
       'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

In [4]:
df.head(20)

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0            0  2008      1           3          4   2003.0        1955   
1            1  2008      1           3          4    754.0         735   
2            2  2008      1           3          4    628.0         620   
3            4  2008      1           3          4   1829.0        1755   
4            5  2008      1           3          4   1940.0        1915   
5            6  2008      1           3          4   1937.0        1830   
6           10  2008      1           3          4    706.0         700   
7           11  2008      1           3          4   1644.0        1510   
8           15  2008      1           3          4   1029.0        1020   
9           16  2008      1           3          4   1452.0        1425   
10          17  2008      1           3          4    754.0         745   
11          18  2008      1           3          4   1323.0        1255   
12          19  2008      1           3          4   1416.0        1325   
13          21  2008      1           3          4   1657.0        1625   
14          22  2008      1           3          4   1900.0        1840   
15          23  2008      1           3          4   1039.0        1030   
16          25  2008      1           3          4   1520.0        1455   
17          26  2008      1           3          4   1422.0        1255   
18          27  2008      1           3          4   1954.0        1925   
19          30  2008      1           3          4   2107.0        1945   

    ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0    2211.0        2225            WN  ...     4.0     8.0          0   
1    1002.0        1000            WN  ...     5.0    10.0          0   
2     804.0         750            WN  ...     3.0    17.0          0   
3    1959.0        1925            WN  ...     3.0    10.0          0   
4    2121.0        2110            WN  ...     4.0    10.0          0   
5    2037.0        1940            WN  ...     3.0     7.0          0   
6     916.0         915            WN  ...     5.0    19.0          0   
7    1845.0        1725            WN  ...     6.0     8.0          0   
8    1021.0        1010            WN  ...     6.0     9.0          0   
9    1640.0        1625            WN  ...     7.0     8.0          0   
10    940.0         955            WN  ...     5.0    16.0          0   
11   1526.0        1510            WN  ...     4.0     9.0          0   
12   1512.0        1435            WN  ...     2.0     5.0          0   
13   1754.0        1735            WN  ...     5.0     5.0          0   
14   1956.0        1950            WN  ...     2.0     5.0          0   
15   1133.0        1140            WN  ...     2.0     5.0          0   
16   1619.0        1605            WN  ...     2.0     7.0          0   
17   1657.0        1610            WN  ...     6.0     6.0          0   
18   2239.0        2235            WN  ...     3.0     7.0          0   
19   2334.0        2230            WN  ...     6.0     7.0          0   

    CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                  N         0           NaN           NaN      NaN   
1                  N         0           NaN           NaN      NaN   
2                  N         0           NaN           NaN      NaN   
3                  N         0           2.0           0.0      0.0   
4                  N         0           NaN           NaN      NaN   
5                  N         0          10.0           0.0      0.0   
6                  N         0           NaN           NaN      NaN   
7                  N         0           8.0           0.0      0.0   
8                  N         0           NaN           NaN      NaN   
9                  N         0           3.0           0.0      0.0   
10                 N         0           NaN           NaN      NaN   
11                 N         0           0

In [5]:
columnes=list(df.columns)

In [6]:
for column in columnes:
    print('* '*6,f'{column} ','* '*6,'\n',df[column].unique()) #per imprimir els valors únics de totes les columnes

* * * * * *  Unnamed: 0  * * * * * *  
 [      0       1       2 ... 7009718 7009726 7009727]
* * * * * *  Year  * * * * * *  
 [2008]
* * * * * *  Month  * * * * * *  
 [ 1  2  3  4  5  6  7  8  9 10 11 12]
* * * * * *  DayofMonth  * * * * * *  
 [ 3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26
 27 28 29 30 31  2  1]
* * * * * *  DayOfWeek  * * * * * *  
 [4 5 6 7 1 2 3]
* * * * * *  DepTime  * * * * * *  
 [2003.  754.  628. ...  502.  503.  425.]
* * * * * *  CRSDepTime  * * * * * *  
 [1955  735  620 ... 2331   48  140]
* * * * * *  ArrTime  * * * * * *  
 [2211. 1002.  804. ...  401.  437.  405.]
* * * * * *  CRSArrTime  * * * * * *  
 [2225 1000  750 ...  305  334  147]
* * * * * *  UniqueCarrier  * * * * * *  
 ['WN' 'XE' 'YV' 'OH' 'OO' 'UA' 'US' 'DL' 'EV' 'F9' 'FL' 'HA' 'MQ' 'NW'
 '9E' 'AA' 'AQ' 'AS' 'B6' 'CO']
* * * * * *  FlightNum  * * * * * *  
 [ 335 3231  448 ... 6869 6870 5272]
* * * * * *  TailNum  * * * * * *  
 ['N712SW' 'N772SW' 'N428WN' ... '

* * * * * *  ArrDelay  * * * * * *  
 [ -14.    2.   14. ...  880. 1279.  991.]
* * * * * *  DepDelay  * * * * * *  
 [  8.  19.  34. ... 956. 810. 925.]
* * * * * *  Origin  * * * * * *  
 ['IAD' 'IND' 'ISP' 'JAN' 'JAX' 'LAS' 'LAX' 'LBB' 'LIT' 'MAF' 'MCI' 'MCO'
 'MDW' 'MHT' 'MSY' 'OAK' 'OKC' 'OMA' 'ONT' 'ORF' 'PBI' 'PDX' 'PHL' 'PHX'
 'PIT' 'PVD' 'RDU' 'RNO' 'RSW' 'SAN' 'SAT' 'SDF' 'SEA' 'SFO' 'SJC' 'SLC'
 'SMF' 'SNA' 'STL' 'TPA' 'TUL' 'TUS' 'ABQ' 'ALB' 'AMA' 'AUS' 'BDL' 'BHM'
 'BNA' 'BOI' 'BUF' 'BUR' 'BWI' 'CLE' 'CMH' 'CRP' 'DAL' 'DEN' 'DTW' 'ELP'
 'FLL' 'GEG' 'HOU' 'HRL' 'ROC' 'ORD' 'EWR' 'SYR' 'IAH' 'CRW' 'FAT' 'COS'
 'MRY' 'LGB' 'BFL' 'EUG' 'ICT' 'MEM' 'BTV' 'MKE' 'LFT' 'BRO' 'PWM' 'MSP'
 'SRQ' 'CLT' 'CVG' 'GSO' 'SHV' 'DCA' 'TYS' 'GSP' 'RIC' 'DFW' 'BGR' 'DAY'
 'GRR' 'CHS' 'CAE' 'TLH' 'XNA' 'GPT' 'VPS' 'LGA' 'ATL' 'MSN' 'SAV' 'BTR'
 'LEX' 'LRD' 'MOB' 'MTJ' 'GRK' 'AEX' 'PNS' 'ABE' 'HSV' 'CHA' 'MFE' 'MLU'
 'DSM' 'MGM' 'AVL' 'LCH' 'BOS' 'MYR' 'CLL' 'DAB' 'ASE' 'ATW' 'BMI' 'CAK'
 'CID' 

 4.130e+02]
* * * * * *  SecurityDelay  * * * * * *  
 [ nan   0.  10.   8.  12.   3.   9.   6.  11.  14.   5.   2.  18.   7.
  22.  17.  20.  13.  16.  34.  19.   4.  15.   1.  45.  63.  48.  36.
  28.  21.  42.  31.  27.  88.  66.  80.  29.  33.  69.  24.  32.  54.
  26.  25.  60.  39.  78.  47.  43.  50.  23.  82.  37.  38. 136. 102.
  30.  40.  61.  46.  68.  41.  52.  58.  85.  35.  67.  57.  49.  44.
  89.  86.  65.  51.  53. 109.  72. 128.  62.  75.  64.  99. 149.  70.
  56.  97.  59. 120. 101. 115. 135. 112.  77.  55.  81.  87. 132. 108.
  74. 126.  98. 142.  92.  83.  76. 131.  71.  73.  91. 392. 175. 139.
 110. 134. 141. 155. 122. 107.  96. 111. 130. 123. 151. 147. 106. 280.
 229. 284. 100. 104. 156.  90. 178. 211.  95.  93. 209. 167. 191. 199.
 105. 140. 114. 124. 170. 150. 186. 121. 216. 113. 117. 357. 169.  79.
 193. 164. 254.]
* * * * * *  LateAircraftDelay  * * * * * *  
 [      nan 3.200e+01 4.700e+01 7.200e+01 1.200e+01 1.600e+01 2.500e+01
 7.000e+00 5.900e+01 6.900e+0

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

In [8]:
df.groupby('Cancelled').count()

Unnamed: 0     Year    Month  DayofMonth  DayOfWeek  DepTime  \
Cancelled                                                                 
0             1936125  1936125  1936125     1936125    1936125  1936125   
1                 633      633      633         633        633      633   

           CRSDepTime  ArrTime  CRSArrTime  UniqueCarrier  ...  Distance  \
Cancelled                                                  ...             
0             1936125  1929648     1936125        1936125  ...   1936125   
1                 633        0         633            633  ...       633   

            TaxiIn  TaxiOut  CancellationCode  Diverted  CarrierDelay  \
Cancelled                                                               
0          1929648  1936125           1936125   1936125       1247488   
1                0      178               633       633             0   

           WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
Cancelled                                                            
0               1247488   1247488        1247488            1247488  
1                     0         0              0                  0  

[2 rows x 29 columns]

In [9]:
df.groupby('Diverted').count()

Unnamed: 0     Year    Month  DayofMonth  DayOfWeek  DepTime  \
Diverted                                                                 
0            1929004  1929004  1929004     1929004    1929004  1929004   
1               7754     7754     7754        7754       7754     7754   

          CRSDepTime  ArrTime  CRSArrTime  UniqueCarrier  ...  Distance  \
Diverted                                                  ...             
0            1929004  1928371     1929004        1929004  ...   1929004   
1               7754     1277        7754           7754  ...      7754   

           TaxiIn  TaxiOut  Cancelled  CancellationCode  CarrierDelay  \
Diverted                                                                
0         1928371  1928549    1929004           1929004       1247488   
1            1277     7754       7754              7754             0   

          WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
Diverted                                                            
0              1247488   1247488        1247488            1247488  
1                    0         0              0                  0  

[2 rows x 29 columns]

Pocs vols cancelats o desviats, més endavant reviso com evoluciona al netejar el dataframe.

In [10]:
del df['Unnamed: 0'] #columna idèntica al índex de posició, és innecesaria i l'esborro

In [11]:
df.drop('Year',axis=1,inplace=True) #No té sentit ja que totes les dades són del 2008
#2a manera de fer-ho, cal picar més codi!

In [12]:
df.head()

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
0      1           3          4   2003.0        1955   2211.0        2225   
1      1           3          4    754.0         735   1002.0        1000   
2      1           3          4    628.0         620    804.0         750   
3      1           3          4   1829.0        1755   1959.0        1925   
4      1           3          4   1940.0        1915   2121.0        2110   

  UniqueCarrier  FlightNum TailNum  ...  TaxiIn  TaxiOut  Cancelled  \
0            WN        335  N712SW  ...     4.0      8.0          0   
1            WN       3231  N772SW  ...     5.0     10.0          0   
2            WN        448  N428WN  ...     3.0     17.0          0   
3            WN       3920  N464WN  ...     3.0     10.0          0   
4            WN        378  N726SW  ...     4.0     10.0          0   

   CancellationCode  Diverted CarrierDelay WeatherDelay  NASDelay  \
0                 N         0          NaN          NaN       NaN   
1                 N         0          NaN          NaN       NaN   
2                 N         0          NaN          NaN       NaN   
3                 N         0          2.0          0.0       0.0   
4                 N         0          NaN          NaN       NaN   

   SecurityDelay  LateAircraftDelay  
0            NaN                NaN  
1            NaN                NaN  
2            NaN                NaN  
3            0.0               32.0  
4            NaN                NaN  

[5 rows x 28 columns]

**Exercici 
2;** 
Fes un informe complet del data set:
* Resumeix estadísticament les columnes d’interès
* Troba quantes dades faltants hi ha per columna
* Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...)
* Taula de les aerolínies amb més endarreriments acumulats
* Quins són els vols més llargs? I els més endarrerits?
Etc.

In [13]:
df.describe() #dona la informació estadística d'interès per a cada columna.

Month    DayofMonth     DayOfWeek       DepTime    CRSDepTime  \
count  1.936758e+06  1.936758e+06  1.936758e+06  1.936758e+06  1.936758e+06   
mean   6.111106e+00  1.575347e+01  3.984827e+00  1.518534e+03  1.467473e+03   
std    3.482546e+00  8.776272e+00  1.995966e+00  4.504853e+02  4.247668e+02   
min    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    3.000000e+00  8.000000e+00  2.000000e+00  1.203000e+03  1.135000e+03   
50%    6.000000e+00  1.600000e+01  4.000000e+00  1.545000e+03  1.510000e+03   
75%    9.000000e+00  2.300000e+01  6.000000e+00  1.900000e+03  1.815000e+03   
max    1.200000e+01  3.100000e+01  7.000000e+00  2.400000e+03  2.359000e+03   

            ArrTime    CRSArrTime     FlightNum  ActualElapsedTime  \
count  1.929648e+06  1.936758e+06  1.936758e+06       1.928371e+06   
mean   1.610141e+03  1.634225e+03  2.184263e+03       1.333059e+02   
std    5.481781e+02  4.646347e+02  1.944702e+03       7.206007e+01   
min    1.000000e+00  0.000000e+00  1.000000e+00       1.400000e+01   
25%    1.316000e+03  1.325000e+03  6.100000e+02       8.000000e+01   
50%    1.715000e+03  1.705000e+03  1.543000e+03       1.160000e+02   
75%    2.030000e+03  2.014000e+03  3.422000e+03       1.650000e+02   
max    2.400000e+03  2.400000e+03  9.742000e+03       1.114000e+03   

       CRSElapsedTime  ...      Distance        TaxiIn       TaxiOut  \
count    1.936560e+06  ...  1.936758e+06  1.929648e+06  1.936303e+06   
mean     1.343027e+02  ...  7.656862e+02  6.812975e+00  1.823220e+01   
std      7.134144e+01  ...  5.744797e+02  5.273595e+00  1.433853e+01   
min     -2.500000e+01  ...  1.100000e+01  0.000000e+00  0.000000e+00   
25%      8.200000e+01  ...  3.380000e+02  4.000000e+00  1.000000e+01   
50%      1.160000e+02  ...  6.060000e+02  6.000000e+00  1.400000e+01   
75%      1.650000e+02  ...  9.980000e+02  8.000000e+00  2.100000e+01   
max      6.600000e+02  ...  4.962000e+03  2.400000e+02  4.220000e+02   

          Cancelled      Diverted  CarrierDelay  WeatherDelay      NASDelay  \
count  1.936758e+06  1.936758e+06  1.247488e+06  1.247488e+06  1.247488e+06   
mean   3.268348e-04  4.003598e-03  1.917940e+01  3.703571e+00  1.502164e+01   
std    1.807562e-02  6.314722e-02  4.354621e+01  2.149290e+01  3.383305e+01   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  2.000000e+00  0.000000e+00  2.000000e+00   
75%    0.000000e+00  0.000000e+00  2.100000e+01  0.000000e+00  1.500000e+01   
max    1.000000e+00  1.000000e+00  2.436000e+03  1.352000e+03  1.357000e+03   

       SecurityDelay  LateAircraftDelay  
count   1.247488e+06       1.247488e+06  
mean    9.013714e-02       2.529647e+01  
std     2.022714e+00       4.205486e+01  
min     0.000000e+00       0.000000e+00  
25%     0.000000e+00       0.000000e+00  
50%     0.000000e+00       8.000000e+00  
75%     0.000000e+00       3.300000e+01  
max     3.920000e+02       1.316000e+03  

[8 rows x 23 columns]

In [14]:
df.isnull().sum() #suma els valors nuls de cada columna i els reporta.

Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

Mostra clarament que on hi manca força informació és en reportar els **minuts de retràs per cada causa**. De 1936758 entrades en fallen 689270 (35,6% del total).
Hi han alguns registres que li falta la **hora d'arribada real**, amb el que no es pot treballar amb les dades de retrasos. Elimino aquests registres (7110) ja que només suposen el 0,37% de les dades totals. 

In [15]:
indexDel=list(df[df['ArrTime'].isnull()].index)#index de les línees que s'han de borrar
df.drop(indexDel, inplace=True)

In [16]:
df.isnull().sum() #suma els valors nuls de cada columna i els reporta.

Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                   0
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   3
ActualElapsedTime      1277
CRSElapsedTime            0
AirTime                1277
ArrDelay               1277
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                    0
TaxiOut                   0
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         682160
WeatherDelay         682160
NASDelay             682160
SecurityDelay        682160
LateAircraftDelay    682160
dtype: int64

Eliminem també les que falten dades de temps de vol o retràs en l'arrivada que manquen i són molt poques respecte el total.

In [17]:
indexDel=list(df[df['AirTime'].isnull()].index)#index de les línees que s'han de borrar
df.drop(indexDel, inplace=True)
df.isnull().sum()

Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                   0
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   3
ActualElapsedTime         0
CRSElapsedTime            0
AirTime                   0
ArrDelay                  0
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                    0
TaxiOut                   0
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         680883
WeatherDelay         680883
NASDelay             680883
SecurityDelay        680883
LateAircraftDelay    680883
dtype: int64

Canviem el valor del número d'avió (TailNum) a 'Unkown' pels 3 registres que no el tenen definit.


In [18]:
df['TailNum'].fillna(value='Unknown',inplace=True)
df.isnull().sum()

Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                   0
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   0
ActualElapsedTime         0
CRSElapsedTime            0
AirTime                   0
ArrDelay                  0
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                    0
TaxiOut                   0
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         680883
WeatherDelay         680883
NASDelay             680883
SecurityDelay        680883
LateAircraftDelay    680883
dtype: int64

In [19]:
df.groupby('Cancelled').count()

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
Cancelled                                                                 
0          1928371     1928371    1928371  1928371     1928371  1928371   

           CRSArrTime  UniqueCarrier  FlightNum  TailNum  ...  Distance  \
Cancelled                                                 ...             
0             1928371        1928371    1928371  1928371  ...   1928371   

            TaxiIn  TaxiOut  CancellationCode  Diverted  CarrierDelay  \
Cancelled                                                               
0          1928371  1928371           1928371   1928371       1247488   

           WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
Cancelled                                                            
0               1247488   1247488        1247488            1247488  

[1 rows x 27 columns]

In [20]:
df.groupby('Diverted').count()

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
Diverted                                                                 
0         1928371     1928371    1928371  1928371     1928371  1928371   

          CRSArrTime  UniqueCarrier  FlightNum  TailNum  ...  Distance  \
Diverted                                                 ...             
0            1928371        1928371    1928371  1928371  ...   1928371   

           TaxiIn  TaxiOut  Cancelled  CancellationCode  CarrierDelay  \
Diverted                                                                
0         1928371  1928371    1928371           1928371       1247488   

          WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
Diverted                                                            
0              1247488   1247488        1247488            1247488  

[1 rows x 27 columns]

Al netejar el dataframe no queden vols cancelats. Es poden borrar totes les columnes referents a la cancelació i desviació de vols.

In [21]:
df.drop(['Cancelled','CancellationCode', 'Diverted'],axis=1,inplace=True)


Finalment generarem un sub-dataframe més petit (2/3 del  total) que contingui les dades de les causes del retràs per quan es vulgui estudiar quines són les causes dels retards.

In [22]:
df_cause=df.copy()
indexDel=list(df_cause[df_cause['NASDelay'].isnull()].index)#index de les línees que s'han de borrar
df_cause.drop(indexDel, inplace=True)
df_cause.isnull().sum()

Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64

In [23]:
#Com ja tenim un dataframe per estudiar especificament els valors de retràs reportats, borrem les olumnes d'aquest dataframe.
df.drop(['CarrierDelay','WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],axis=1,inplace=True)
df.isnull().sum()

Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
dtype: int64

### Crear Columnes noves

In [24]:
df['AverageSpeed']=df['Distance']/(df['AirTime']/60)#Dividim entre 60 perquè són minuts i volem per hora

In [25]:
df['AverageSpeed'].head(20)

0     418.965517
1     430.088496
2     406.578947
3     401.298701
4     474.482759
5     415.043478
6     468.679245
7     464.299065
8     262.702703
9     419.436620
10    435.804878
11    457.090909
12    269.387755
13    280.851064
14    269.387755
15    280.851064
16    264.000000
17    458.601399
18    423.096774
19    435.223881
Name: AverageSpeed, dtype: float64

In [26]:
df['Delayed']=df['ArrTime']>df['CRSArrTime'] #Generem columna de si hi ha retràs (=True) o no (False)

In [27]:
df['UniqueCarrier'].value_counts().head(3)

WN    376201
AA    190910
MQ    141223
Name: UniqueCarrier, dtype: int64

Les companyies amb més vols són WN (Southwest Airlines), AA (American Airlines) i MQ(American Eagle Airlines).

In [28]:
df_Carrier=df.groupby('UniqueCarrier')

In [29]:
df2=pd.DataFrame(df_Carrier['Delayed'].value_counts(True))
df2.xs(True, level='Delayed')[df2.xs(True, level='Delayed')['Delayed']==df2.xs(True, level='Delayed')['Delayed'].max()]

Delayed
UniqueCarrier         
HA             0.92773

In [30]:
df2.xs(True, level='Delayed')[df2.xs(True, level='Delayed')['Delayed']==df2.xs(True, level='Delayed')['Delayed'].min()]

Delayed
UniqueCarrier          
CO             0.780399

**La companyia que proporcionalment es retrassa més és HA** (Hawaiian Airlines) amb un 92,8% dels seus vols amb retràs.
**La companyia que proporcionalment es retrassa menys és CO** (Continental Express) amb un 78,0% dels seus vols amb retràs.

In [31]:
df2.xs(True, level='Delayed').sort_values('Delayed').iloc[0:3]# Les tres aerolinies més puntuals

Delayed
UniqueCarrier          
CO             0.780399
B6             0.785890
US             0.808452

In [32]:
df2.xs(True, level='Delayed').sort_values('Delayed',ascending=False).iloc[0:3]# Les tres aerolinies més impuntuals

Delayed
UniqueCarrier          
HA             0.927730
YV             0.906169
MQ             0.905185

In [33]:
df[df['UniqueCarrier'].isin(['HA', 'YV','MQ'])] #Data Frame amb les tres companyies més impuntuals.

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
40993        1           1          2   2036.0        1741   2128.0   
40994        1           1          2   1447.0        1230   1553.0   
40995        1           1          2   1211.0        1153   1323.0   
40996        1           1          2    930.0         900   1027.0   
40997        1           1          2   1346.0        1337   1448.0   
...        ...         ...        ...      ...         ...      ...   
1879355     12          18          4   1548.0        1535   1737.0   
1879356     12          19          5   1608.0        1535   1920.0   
1879357     12          21          7   1555.0        1535   1801.0   
1879358     12          22          1   1611.0        1535   1804.0   
1879359     12          24          3   1603.0        1535   1752.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ...  AirTime  ArrDelay  \
40993          1856            YV       7138  N508MJ  ...    101.0     152.0   
40994          1347            YV       7263  N650ML  ...    114.0     126.0   
40995          1256            YV       7069  N455YV  ...     36.0      27.0   
40996          1011            YV       7074  N455YV  ...     39.0      16.0   
40997          1440            YV       7098  N434YV  ...     38.0       8.0   
...             ...           ...        ...     ...  ...      ...       ...   
1879355        1730            MQ       4919  N712MQ  ...     87.0       7.0   
1879356        1730            MQ       4919  N715MQ  ...     93.0     110.0   
1879357        1730            MQ       4919  N710MQ  ...     93.0      31.0   
1879358        1730            MQ       4919  N724MQ  ...     79.0      34.0   
1879359        1730            MQ       4919  N702MQ  ...     87.0      22.0   

         DepDelay  Origin  Dest Distance TaxiIn  TaxiOut  AverageSpeed  \
40993       175.0     ABE   ORD      654    5.0      6.0    388.514851   
40994       137.0     ABE   ORD      654    5.0      7.0    344.210526   
40995        18.0     ASE   DEN      125    6.0     30.0    208.333333   
40996        30.0     ASE   DEN      125    8.0     10.0    192.307692   
40997         9.0     ASE   DEN      125    8.0     16.0    197.368421   
...           ...     ...   ...      ...    ...      ...           ...   
1879355      13.0     LGA   CMH      478    3.0     19.0    329.655172   
1879356      33.0     LGA   CMH      478    7.0     92.0    308.387097   
1879357      20.0     LGA   CMH      478    8.0     25.0    308.387097   
1879358      36.0     LGA   CMH      478    3.0     31.0    363.037975   
1879359      28.0     LGA   CMH      478    4.0     18.0    329.655172   

         Delayed  
40993       True  
40994       True  
40995       True  
40996       True  
40997       True  
...          ...  
1879355     True  
1879356     True  
1879357     True  
1879358     True  
1879359     True  

[215464 rows x 22 columns]

In [34]:
listDist=df['Distance'].unique()

listDist.sort()
print(listDist[-5:])
 #els 10 vols més llargs



[4184 4213 4243 4502 4962]


In [35]:
df_Dist=pd.DataFrame(df.groupby('Distance')['FlightNum'].value_counts()).tail(10) 
df_Dist # Són els 10 # de vols amb les distàncies més llargues. També es detalla quants de cada un es van registrar a la última columna.

FlightNum
Distance FlightNum           
4243     72                32
         1410               1
4502     1273             180
         1282             107
         851               67
         1561              59
         850               25
         1560              19
4962     15               265
         14                57

A continuació es detallen els trajectes del 3 vols més llargs, tots tres surten o arriben a HNL = Honolulu

In [36]:
df[df['Distance']==4962].groupby('Origin')['Dest'].value_counts()          

Origin  Dest
EWR     HNL     265
HNL     EWR      57
Name: Dest, dtype: int64

In [37]:
df[df['Distance']==4502].groupby('Origin')['Dest'].value_counts()  

Origin  Dest
ATL     HNL     306
HNL     ATL     151
Name: Dest, dtype: int64

In [38]:
df[df['Distance']==4243].groupby('Origin')['Dest'].value_counts()  

Origin  Dest
HNL     ORD      91
ORD     HNL     206
Name: Dest, dtype: int64

In [39]:
df.sort_values('ArrDelay', ascending=False)[['ArrDelay','FlightNum','UniqueCarrier']].head() #Els 5 vols amb més retard
#Es detalla el retràs en la columna 'ArrDelay', el número de vol i la companyia

ArrDelay  FlightNum UniqueCarrier
322516     2461.0        808            NW
686014     2453.0       1699            NW
839306     1951.0       1107            NW
1009553    1707.0       3538            MQ
1881639    1655.0        357            NW

**Exercici 
3:**
Exporta el data set net i amb les noves columnes a Excel.

In [48]:
df.to_excel('2008_flights_in_USA.xlsx',index=False)

ValueError: This sheet is too large! Your sheet size is: 1928371, 22 Max sheet size is: 1048576, 16384

In [40]:
df_1apart = df.iloc[:1000000,:].copy()#El dataset és massa gran per ser guardat a Excel, el dividim en 2.
df_2apart = df.iloc[1000000:, :].copy()

In [46]:
df_2apart.to_excel('2008_flights_in_USA__2apart.xlsx',index=False, engine='xlsxwriter')

Exception ignored in: <function ZipFile.__del__ at 0x0000027915F7D9D0>
Traceback (most recent call last):
  File "C:\Users\march\anaconda3\envs\notebook\lib\zipfile.py", line 1807, in __del__
  File "C:\Users\march\anaconda3\envs\notebook\lib\zipfile.py", line 1824, in close
ValueError: seek of closed file


In [47]:
df_1apart.to_excel('2008_flights_in_USA__1apart.xlsx',index=False, engine='xlsxwriter')

He penjat els dos arxius excel dintre d'un arxiu zip dividit en 8 perquè a Git-Hub no deixa guardar arxius superiors a 20Mb